### DISCUSSION TITLE
[Proposal] Andoid/iOS Companion App to Sent SMS

### DISCUSSION THREAD

**Discussion by daltonfury42** [SimplQ/simplQ-frontend#496] [Ideas]
#### Usecase

Sending SMS to the people in the queue when they are added to the queue, and when their turn is up, will be a value adding feature. 

We had explored paid SMS services and have realised:

1. SMS would incur costs for us, and will be come an expensive resource. 
2. We would then have to limit the usage, or make it a paid feature, but we want to keep this free.
3. Global SMS delivery will add to the expense.

#### Proposal

Shifting responsibility of sending SMS to the queue manager / client side would mean that we won't incur costs, won't have to protect/meter/bill this paid resource and also not have to worry about how to deliver SMS globally.

Idea is to have a companion app, that the queue manager can optionally install on his mobile phone, which will allow him to optionally send out SMS from his phone in the background. 

#### The Flow

1. As usual admin creates and manages a normal queue. 
2. On the admin side panel, there will be a "Configure SMS" button. If the admin wishes to set up SMS Notifications for the queue, clicking on it instruct him to:

> Step 1: Go to App Store and install the SimplQ Companion App.
> Step 2: Link the app with the queue by entering the passcode: <random-number>
> Congratulations! Your phone is successfully configured to send SMS

When a user is added to the queue, he will then get a SMS with the token status page link:

```
Hi <token-user-name>,
You have been added to <queue-name>. Your token number is <XX>. You can know your status like by visiting <token-status-page-link>.

Thanks for using simplq.me
```

When notify button is pressed by the queue manager, the app sends an SMS from the configured phone.

```
Hi <token-user-name>,
Hooray! Your turn is up! You have been notified by the queue manager.

Thanks for using simplq.me
```

#### The App

We will have to build a very simple app for both iOS and Android, with a single screen:
```
============================
 
<SimplQ Logo>

Passcode: ________________
      {Connect Button}

============================
```

Once the app is linked with the queue, it will then receive push notifications from the backend on events, and the app will then send SMS from the device.

#### Future Enhancements

1. We can configure which events (user_joined, user_notified etc) should trigger SMS.
2. We can in future, slowly expand the app to let the user also create and manage queues, if we want to.
3. Passcode can be replaced by a [google login for registered users](https://developers.google.com/identity/sign-in/android/sign-in).

**Comment by maaverik** [comment]
This is a clever idea for the cost issue, but I'm not too optimistic that anyone would really use this. Firstly, it's money from their own pockets and secondly, they'd be letting an app send messages from their phones which is very very rare for apps. I've only ever seen this during UPI account verification. Maybe one day when simplq has good reputation, people would trust it enough to allow sending messages, but the cost for them is still an issue.

What are your thoughts on a whatsapp bot instead. I'm not sure, but I think it's currently free for use. This way, we don't have any cost issue (if it is free), we take care of notifications, and there would be no need for an additional app, at least for now.

**Comment by daltonfury42** [comment]
**Whatsapp vs SMS:** SMS is more universal, everyone who has a mobile number would receive text messages. Imagine a clinic in Kerala wants to use SimplQ, and the receptionist would add patients to queue manually, and then telling them, "Uncle, your token number is 18, you will also recieve receive an SMS on your mobile with token number." might cover 90% patients, while a lot of people might not have mobile data, or Whatsapp etc. 

```
Oh, this number doesn't have whatsapp.
Oh, I don't have internet.
Do you have whatsapp installed? Is this your whatsapp number?
```
we want to avoid all these conversations.

**Cost**: For a business user, cost is not an issue, if something adds value. If we pitch to a hospital owner and ask him if he wants a whatsapp message to be send, which we can do for free, vs we can set up SMS for you, but you have to pay the SMS charge, he will go for the SMS option is what I think.

I randomly checked online, and Jio lets you send 9000 SMS for Rs. 499 for three months. This is negligible cost for a business user.

But we could definitely think of integrating Whatsapp, @maaverik would you like to take this up and explore how to integrate WA in a separate thread?

**Comment by sajmalyousef** [comment]
![DAA8B44A-EFB2-4719-8CBA-5A7753B94EA3](https://user-images.githubusercontent.com/13188906/98463517-b13f3580-21e1-11eb-83a9-1e163e2d85e7.jpeg)

I’ve assumed people less than 10k income don’t have smartphones which may not be the case often.

The number maybe off by a bit, but is this a big deal , assuming 
1. only tech savvy people would initially adopt our app
2. Queue status page is built on the assumption that people would have smartphones and data.

**Comment by daltonfury42** [comment]
@sajmalyousef what's the conclusion? Should we have SMS or not?

We should definitely integrate Whatsapp alerts if that's possible, but it is not straightforward to do that for our use case from my research.

https://developers.facebook.com/docs/whatsapp/faq

> Can I send any message to any user at any time?
> No.
> 
> If a user contacts an enterprise, the enterprise can respond with any type of message in the next 24 hours. This type of message is free.
> 
> But if the enterprise is contacting a user before the user sends a message or after more than 24 hours have passed, the enterprise can only send a message template. This is a paid notification.

We might be able to integrate Whatsapp in a different usecase, but it won't work for us here for free

**Comment by daltonfury42** [comment]
I have slowly started working on an Android app prototype. I have no experience in writing android apps, just trying to make it work. So it would require lot of refactoring later on: https://github.com/SimplQ/simplQ-android

Some other related discussions here:  https://github.com/SimplQ/simplQ-backend/issues/101

**Reply by mmendiratta** [reply]
So is the goal of the android app to be used by the queue manager to be able to send out some type of notifications from their android device? Maybe more directly, who is the main user of the app?

**Reply by daltonfury42** [reply]
Yes, right. The main user of the app is the queue manager who will install it, run the app in the background / put his phone away. The app sends SMS from the phone in the background.
